kbs/src/api/src/policy_engine/opa/mod.rs

```diff
@@ -2,43 +2,25 @@
 // Licensed under the Apache License, Version 2.0, see LICENSE for details.
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::policy_engine::PolicyEngineInterface;
-use anyhow::{anyhow, Result};
+use crate::policy_engine::{PolicyEngineInterface, ResourcePolicyError};
 use async_trait::async_trait;
 use base64::Engine;
-use serde_json::Value;
-use std::ffi::CStr;
 use std::fs;
-use std::os::raw::c_char;
 use std::path::PathBuf;
 
-// Link import cgo function
-#[link(name = "cgo")]
-extern "C" {
-    pub fn evaluateGo(policy: GoString, data: GoString, input: GoString) -> *mut c_char;
-}
-
-/// String structure passed into cgo
-#[derive(Debug)]
-#[repr(C)]
-pub struct GoString {
-    pub p: *const c_char,
-    pub n: isize,
-}
-
 #[derive(Debug, Clone)]
 pub struct Opa {
     policy_path: PathBuf,
 }
 
 impl Opa {
-    pub fn new(policy_path: PathBuf) -> Result<Self> {
+    pub fn new(policy_path: PathBuf) -> Result<Self, ResourcePolicyError> {
+        std::fs::create_dir_all(policy_path.parent().unwrap())
+            .map_err(|_| ResourcePolicyError::IOError)?;
+
         if !policy_path.as_path().exists() {
-            if !policy_path.as_path().parent().unwrap().exists() {
-                std::fs::create_dir_all(policy_path.parent().unwrap())?;
-            }
             let policy = std::include_str!("default_policy.rego").to_string();
-            fs::write(&policy_path, policy)?;
+            fs::write(&policy_path, policy).map_err(|_| ResourcePolicyError::IOError)?;
```

Comment:
If we add a `from` directive for the source errors, we don't need the explicit mapping (can just use `fs::write(&policy_path, policy)?;`) and keep the context of the original error.

```rust
#[derive(Error, Debug, PartialEq)]
pub enum ResourcePolicyError {
    #[error("io error")]
    IoError(#[from] io::Error),
    ...
}
```

Reply:
Ok, `io::Error` does not implement `PartialEq`, which breaks the asserts below. This is generally the reason why I don't have context in any of the errors, but I agree it would be better to carry more info so let me see if there is a good workaround.

Reply:
ah, yes that's a bit unergonomic, but I wouldn't discard the source error because of this. there is a couple of ways, but the simplest would be to drop the `PartialEq` derive and do assertions via pattern matching or a helper macro (assert_matches):

```rust
use std::io;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum MyError {
    #[error("fail")]
    IoFail(#[from] io::Error),
    #[error("sth else")]
    SthElse,
}

fn fails() -> Result<(), MyError> {
    let err = Err(io::Error::new(io::ErrorKind::NotFound, "fail"));
    err?;
    Ok(())
}

#[test]
fn test_fail() {
    let MyError::IoFail(_) = fails().unwrap_err() else {
        panic!("Expected an IoFail error");
    };
}
```

Reply:
Ok, turns out the `matches!` macros have to match against a specific pattern so I made a `compare_errors` function instead with a big match statement in it. Still slightly weird but this is the best way I can think of. Also, it turns out that thiserror doesn't support multiple variants coming from the same source. Since regorus uses anyhow for all its errors this is slightly limiting. I opted to include the error just for the evaluate case (since i can only do one). I guess I could make one variant for all the regorus errors, but this seems fine.